pystac/extensions/timestamps.py

```diff
@@ -0,0 +1,161 @@
+import pystac
+from pystac import Extensions
+from pystac.extensions.base import (ExtendedObject, ExtensionDefinition, ItemExtension)
+from pystac.item import Item
+from pystac.utils import datetime_to_str, str_to_datetime
+
+
+class TimestampsItemExt(ItemExtension):
+    """TimestampsItemExt is the extension of an Item in that
+    allows to specify additional timestamps for assets and metadata.
+
+    Args:
+        item (Item): The item to be extended.
+
+    Attributes:
+        item (Item): The Item that is being extended.
+
+    Note:
+        Using TimestampsItemExt to directly wrap an item will add the 'timestamps'
+        extension ID to the item's stac_extensions.
+    """
+    def __init__(self, item):
+        if item.stac_extensions is None:
+            item.stac_extensions = [Extensions.TIMESTAMPS]
+        elif Extensions.TIMESTAMPS not in item.stac_extensions:
+            item.stac_extensions.append(Extensions.TIMESTAMPS)
+
+        self.item = item
+
+    @classmethod
+    def from_item(cls, item):
+        return cls(item)
+
+    @classmethod
+    def _object_links(cls):
+        return []
+
+    def apply(self, published=None, expires=None, unpublished=None):
+        """Applies timestamps extension properties to the extended Item.
+
+        Args:
+            published (datetime or None): Date and time the corresponding data
+                (see below) was published the first time.
+            expires (datetime or None): Date and time the corresponding data
+                (see below) expires (is not valid any longer).
+            unpublished (datetime or None): Date and time the corresponding data
+                (see below) was unpublished.
+        """
+        if published is None and expires is None and unpublished is None:
+            raise pystac.STACError("timestamps extension needs at least one property value.")
+
+        self.published = published
+        self.expires = expires
+        self.unpublished = unpublished
+
+    def _timestamp_getter(self, key, asset=None):
+        if asset is not None and key in asset.properties:
+            timestamp_str = asset.properties.get(key)
+        else:
+            timestamp_str = self.item.properties.get(key)
+
+        timestamp = None
+        if timestamp_str is not None:
+            timestamp = str_to_datetime(timestamp_str)
+
+        return timestamp
+
+    def _timestamp_setter(self, timestamp, key, asset=None):
+        if timestamp is None:
+            self.item.properties[key] = timestamp
+        else:
+            timestamp_str = datetime_to_str(timestamp)
+            if asset is not None:
+                asset.properties[key] = timestamp_str
+            else:
+                self.item.properties[key] = timestamp_str
+
+    @property
+    def published(self):
```

Comment:
I'd suggest putting a docstring that reads "get or sets ..." that includes the property description from the STAC spec, e.g. "Gets or sets the date and time the corresponding data (see below) was published the first time."

The convention is to doc both the property getter and setter in the @property getter: https://stackoverflow.com/questions/16025462/what-is-the-right-way-to-put-a-docstring-on-python-property